bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/BasicChannelTypeProvider.java

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.miio.internal.basic;
+
+import static org.openhab.binding.miio.internal.MiIoBindingConstants.BINDING_ID;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.thing.type.ChannelType;
+import org.openhab.core.thing.type.ChannelTypeBuilder;
+import org.openhab.core.thing.type.ChannelTypeProvider;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.openhab.core.thing.type.StateChannelTypeBuilder;
+import org.openhab.core.types.StateDescriptionFragmentBuilder;
+import org.openhab.core.types.StateOption;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provide channelTypes for Mi IO Basic devices
+ *
+ * @author Marcel Verpaalen - Initial contribution
+ */
+@Component(service = { ChannelTypeProvider.class, BasicChannelTypeProvider.class })
+@NonNullByDefault
+public class BasicChannelTypeProvider implements ChannelTypeProvider {
+    private final Map<String, ChannelType> channelTypes = new ConcurrentHashMap<>();
+    private final Logger logger = LoggerFactory.getLogger(BasicChannelTypeProvider.class);
+
+    @Override
+    public Collection<ChannelType> getChannelTypes(@Nullable Locale locale) {
+        return channelTypes.values();
+    }
+
+    @Override
+    public @Nullable ChannelType getChannelType(ChannelTypeUID channelTypeUID, @Nullable Locale locale) {
+        if (channelTypes.containsKey(channelTypeUID.getAsString())) {
+            return channelTypes.get(channelTypeUID.getAsString());
+        }
+        return null;
+    }
+
+    public void addChannelType(MiIoBasicChannel miChannel, String model) {
```

Comment:
I noticed in my own bindings that my channels would not show up in the UI if the channel types weren't available in the ChannelTypeProvider before the thing handler was started. Is there something special about your setup here or did you find some kind of workaround to prevent that issue?

Reply:
I have not experienced that. I'll do some specific testing on that.
It very well can be the case, but as the basicHandler changes the type from generic to basic after it did the identification it would have satisfied that requirement. But... I would need to check the code if it still does that when the type has already changed, I guess than indeed it may be an issue.
I wonder how other bindings are doing this, as in many cases the channel definition is based on something that is loaded of the device.
How did you solve it?

Reply:
I ended encoding all of the relevant channel information into the channel type name itself. Wish I didn't have to do that and that there was some way around it so that's why I was asking if you are experiencing the same issue.

Reply:
Strange. I don't see the behavior. If I change the model info on an existing thing, the new channels appear in the mainUI almost instant. No reloading of the page (I go from the code page to the channel page and the new channels are there)
No specific actions I take to handle the reloading the handler.

Reply:
Well if it works for you then that is good enough to me. I'll sort out my own setup some other time.

Reply:
@cpmeister could you point out the code you've used so we can have a look (cc @Markinus)
We're having the same issue with a binding where the channels show up in the console but not in the UI on OH3, and it appears to be related to channel ChannelType. All works well in OH2.5
Thx!

Reply:
As I mentioned before, I had to implement the ChannelTypeProvider such that the ChannelType information could be derived from the ChannelTypeUID without having a handler inject ChannelType definitions into the provider ahead of time.

https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/CharacteristicChannelTypeProvider.java

@fastlorenzo What binding are you having trouble with?

Reply:
Thanks for the quick reply!

I ended up fixing it by checking in the ChannelTypeRegistry if the ChannelType exists before dynamically creating the channels.

This is the Carnet binding (interface with Audi/VW API), so all channels are taken from the car info, and are different even with the same car model (depending on options for example)

https://github.com/markus7017/openhab-addons/pull/14/files